WcaOnRails/lib/advancement_condition.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class AdvancementCondition
+  include ActiveModel::Validations
+
+  attr_accessor :level
+  validates :level, numericality: { only_integer: true }
+
+  def initialize(level)
+    self.level = level
+  end
+
+  def to_wcif
+    { "type" => self.class.wcif_type, "level" => self.level }
+  end
+
+  def ==(other)
+    other.class == self.class && other.to_wcif == self.to_wcif
+  end
+
+  def hash
+    self.to_wcif.hash
+  end
+
+  def self.wcif_type_to_class
+    @@wcif_type_to_class ||= AdvancementCondition.subclasses.map { |cls| [cls.wcif_type, cls] }.to_h
```

Comment:
I'm thinking a bit beyond this PR and assuming a mapping like this will exist for all/most classes/types. If that's not true, my argument is considerable weaker.

I thought I saw other code also dealing with this mapping in the bigger PR, but I don't see it here.